notebook/services/kernels/kernelmanager.py

```diff
@@ -99,6 +101,47 @@ def shutdown_kernel(self, kernel_id, now=False):
         self._check_kernel_id(kernel_id)
         super(MappingKernelManager, self).shutdown_kernel(kernel_id, now=now)
 
+    def restart_kernel(self, kernel_id):
+        """Restart a kernel by kernel_id"""
+        self._check_kernel_id(kernel_id)
+        super(MappingKernelManager, self).restart_kernel(kernel_id)
+        kernel = self.get_kernel(kernel_id)
```

Comment:
Okay, I *think* I understand now.  This [line](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/manager.py#L336) both ensures that the previous process is closed, and was what was causing the race condition I was seeing.  I would sometimes get the REST API response prior to the previous kernel process being shut down.  So, I concur with this change and https://github.com/jupyter/jupyter-js-services/pull/42.